src/RRule.php

```diff
@@ -0,0 +1,309 @@
+
+    public function getStart(): ?DateTimeInterface
+    {
+        return $this->rrule->getStartDate();
+    }
+
+    /**
+     * Set the end time of this frequency
+     *
+     * The given datetime will be cloned and microseconds removed since iCalendar datetimes only work to the second.
+     *
+     * @param DateTimeInterface $end
+     *
+     * @return $this
+     */
+    public function endAt(DateTimeInterface $end): self
+    {
+        $end = clone $end;
+        $end->setTime($end->format('H'), $end->format('i'), $end->format('s'));
+
+        $this->rrule->setEndDate($end);
+
+        return $this;
+    }
+
+    public function getEnd(): ?DateTimeInterface
+    {
+        return $this->rrule->getEndDate() ?? $this->rrule->getUntil();
+    }
+
+    /**
+     * Get the frequency of this rule
+     *
+     * @return string
+     */
+    public function getFrequency(): string
+    {
+        return $this->frequency;
+    }
+
+    /**
+     * Get a set of recurrences relative to the given time and bounded to the configured generator's limit
+     *
+     * @param DateTimeInterface $dateTime
+     * @param int $limit Limit the recurrences to be generated to the given value
+     * @param bool $include Whether to include the passed time in the result set
+     *
+     * @return Generator
+     */
+    public function getNextRecurrences(
+        DateTimeInterface $dateTime,
+        int $limit = self::DEFAULT_LIMIT,
+        bool $include = true
+    ): Generator {
+        $resetTransformerConfig = function (int $limit = self::DEFAULT_LIMIT): void {
+            $this->transformerConfig->setVirtualLimit($limit);
+            $this->transformer->setConfig($this->transformerConfig);
+        };
+
+        if ($limit > self::DEFAULT_LIMIT) {
+            $resetTransformerConfig($limit);
+        }
+
+        $constraint = new AfterConstraint($dateTime, $include);
+        if (! $this->rrule->repeatsIndefinitely()) {
+            // When accessing this method externally (not by using `getNextDue()`), the transformer may
+            // generate recurrences beyond the configured end time.
+            $constraint = new BetweenConstraint($dateTime, $this->getEnd(), $include);
+        }
+
+        // Setting the start date to a date time smaller than now causes the underlying library
+        // not to generate any recurrences when using the regular frequencies such as `MINUTELY` etc.
+        // and the `$countConstraintFailures` is set to true. We need also to tell the transformer
+        // not to count the recurrences that fail the constraint's test!
+        $recurrences = $this->transformer->transform($this->rrule, $constraint, false);
+        foreach ($recurrences as $recurrence) {
+            yield $recurrence->getStart();
+        }
+
+        if ($limit > self::DEFAULT_LIMIT) {
+            $resetTransformerConfig();
+        }
+    }
+
+    #[\ReturnTypeWillChange]
+    public function jsonSerialize()
+    {
+        return [
```

Comment:
JSON == string

Reply:
No you return the data to be serialized.

Reply:
Other implementations are also wrong :D.

Reply:
:scream: -.-

Reply:
What do we do with `fromJson` then? Does it only accept arrays?

Reply:
And this should define `: array` then.

Reply:
`fromJson` should expect a JSON `string`. No validation inside. Just like `serialize` and `unserialize`.

Reply:
> And this should define `: array` then.

Should be `mixed` to comply with the interface, right?

Reply:
array is a subset of mixed, phpstorm doesn't complain at least and since we have lint tests with PHP 8.2 the tests wouldn't succeed if not.

Reply:
Ok, sounds good!

Reply:
@yhabteab Please add `: array`.